src/RoslynInsertionTool/RoslynInsertionTool/RoslynInsertionTool.VisualStudioTeamServices.cs

```diff
@@ -96,14 +81,34 @@ where buildClient.GetArtifactsAsync(build.Project.Id, build.Id, cancellationToke
         }
 
         private static async Task<Build> GetLatestBuildAsync(CancellationToken cancellationToken)
+        {
+            var buildClient = ProjectCollection.GetClient<BuildHttpClient>();
+            var definitions = await buildClient.GetDefinitionsAsync(project: Options.TFSProjectName, name: Options.BuildQueueName);
+            var builds = await GetBuildsFromTFSAsync(buildClient, definitions, cancellationToken, resultFilter: null);
+
+            return (from build in builds
+                    orderby build.FinishTime descending
+                    select build).FirstOrDefault();
+        }
+
+        private static async Task<Build> GetLatestPassedBuildAsync(CancellationToken cancellationToken)
         {
             // ********************* Verify Build Passed *****************************
             cancellationToken.ThrowIfCancellationRequested();
             Build newestBuild = null;
             Log.Info($"Get Latest Passed Build");
             try
             {
-                newestBuild = await GetLatestPassedBuildAsync(cancellationToken);
+                Log.Trace($"Getting latest passing build from {Options.TFSProjectName} where name is {Options.BuildQueueName}");
+                var buildClient = ProjectCollection.GetClient<BuildHttpClient>();
+                var definitions = await buildClient.GetDefinitionsAsync(project: Options.TFSProjectName, name: Options.BuildQueueName);
+                var builds = await GetBuildsFromTFSAsync(buildClient, definitions, cancellationToken, BuildResult.Succeeded);
+
+                // Get the latest build with valid artifacts.
+                newestBuild = (from build in builds
+                        where buildClient.GetArtifactsAsync(build.Project.Id, build.Id, cancellationToken).Result.Any(a => a.Resource != null && a.Resource.Data != null && a.Resource.Data.Contains(Options.BuildDropPath))
```

Comment:
Note we still need this check in for the "latest" build -- this filters out prodcon builds that we can't insert either.

Reply:
@jasonmalinowski I think I've fixed this.  Unfortunately LINQ doesn't mix well with async so it's uglier now, but we're also not accessing the `Result` of an async method's return value anymore either.